hathor/indexes/memory_deps_index.py

```diff
@@ -0,0 +1,170 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import TYPE_CHECKING, Dict, FrozenSet, Iterator, List, Set, Tuple
+
+from structlog import get_logger
+
+from hathor.indexes.deps_index import DepsIndex, get_validation_state
+from hathor.transaction import BaseTransaction
+from hathor.transaction.transaction_metadata import ValidationState
+from hathor.util import not_none
+
+if TYPE_CHECKING:  # pragma: no cover
+    from hathor.transaction.storage import TransactionStorage
+
+logger = get_logger()
+
+
+class _DirDepValue(Dict[bytes, ValidationState]):
+    """This class is used to add a handy method to values on dependency indexes."""
+
+    def is_ready(self) -> bool:
+        """True if all deps' validation are fully connected."""
+        return all(val.is_fully_connected() for val in self.values())
+
+
+class MemoryDepsIndex(DepsIndex):
+    def __init__(self):
+        self.log = logger.new()
+
+        # Direct and reverse dependency mapping (i.e. needs and needed by)
+        self._dir_dep_index: Dict[bytes, _DirDepValue] = {}
+        self._rev_dep_index: Dict[bytes, Set[bytes]] = {}
+        self._txs_with_deps_ready: Set[bytes] = set()
+
+        # Needed txs (key: tx missing, value: requested by)
+        self._needed_txs_index: Dict[bytes, Tuple[int, bytes]] = {}
+
+    def update_validation(self, tx: bytes, validation: ValidationState) -> None:
+        if tx in self._rev_dep_index:
+            self._update_validation(tx, validation)
+
+    def _update_validation(self, tx: bytes, validation: ValidationState) -> None:
+        """Propagate the new validation state of a given dep."""
+        for cousin in self._rev_dep_index[tx].copy():
+            deps = self._dir_dep_index[cousin]
+            # XXX: this check serves to avoid calling is_ready() when nothing changed
+            if deps[tx] != validation:
+                deps[tx] = validation
+                if deps.is_ready():
+                    self._del_from_deps_index(cousin)
+                    self._txs_with_deps_ready.add(cousin)
+
+    def update(self, tx: BaseTransaction, partial: bool = True) -> None:
+        assert tx.hash is not None
+        assert tx.storage is not None
+        validation = get_validation_state(tx.storage, tx.hash)
+        if validation.is_fully_connected():
+            self._del_from_deps_index(tx.hash)
+        elif not partial:
+            raise ValueError('partial=False will only accept fully connected transactions')
+        else:
+            self._add_deps(tx)
+            self._add_needed(tx)
+
+    def _add_deps(self, tx: BaseTransaction) -> None:
+        assert tx.hash is not None
+        assert tx.storage is not None
+        # deps are immutable for a given hash
+        _deps = _DirDepValue((dep, get_validation_state(tx.storage, dep)) for dep in tx.get_all_dependencies())
```

Comment:
It seems we are caching the state of the transactions. Isn't it better to just hit the database and avoid dirty values?

Reply:
I think that is the case. Removing the "direct dependencies+validation_state" portion of this index could be a very good clean up. But it might be best to validate this on top of sync-v2 first.

Reply:
I've removed those, it is significantly simpler now.

Reply:
Great! Thanks!